src/segment.hpp

```diff
@@ -11,9 +14,60 @@
 #include <mechinfo.hpp>
 #include <point.hpp>
 #include <util/make_unique.hpp>
+#include <util/maputil.hpp>
 
 namespace arb {
 
+// Mechanism information attached to a segment.
+
+struct mechanism_desc {
+    struct field_proxy {
+        mechanism_desc* m;
+        std::string key;
+
+        field_proxy& operator=(double v) {
+            m->set(key, v);
+            return *this;
+        }
+
+        operator double() const {
+            return m->get(key);
+        }
+    };
+
+    // implicit
+    mechanism_desc(std::string name): name_(std::move(name)) {}
+    mechanism_desc(const char* name): name_(name) {}
+
+    mechanism_desc& set(std::string key, double value) {
+        param_[key] = value;
+        return *this;
+    }
+
+    double operator[](const std::string& key) const {
+        return get(key);
+    }
+
+    field_proxy operator[](const std::string& key) {
+        return {this, key};
+    }
+
+    double get(const std::string& key) const {
+        auto optv = util::value_by_key(param_, key);
+        return optv? *optv: throw std::out_of_range("no field "+key+" set");
+    }
+
+    const std::unordered_map<std::string, double>& values() const {
+        return param_;
+    }
+
+    const std::string& name() const { return name_; }
+
+private:
+    std::string name_;
```

Comment:
Is `name_` the name of the mechanism? Maybe `mech_name_`.

Reply:
I thought this one was fairly self-explanatory; `name()` provides view to `name_` which is the name field in a `mechanism_desc` etc. I'd prefer to keep the short member name.